tests/integration/helpers.py

```diff
@@ -372,3 +373,142 @@ def cluster_name(unit: Unit, model_name: str) -> str:
+    """Returns the read-only-endpoints from the relation data.
+
+    Args:
+        relation_data: The dictionary that contains the info
+    Returns:
+        a set that contains the read-only-endpoints
+    """
+    related_units = relation_data[0]["related-units"]
+    roe = set()
+    for _, r_data in related_units.items():
+        assert "data" in r_data
+        data = r_data["data"]["data"]
+
+        try:
+            j_data = json.loads(data)
+            if "read-only-endpoints" in j_data:
+                read_only_endpoints = j_data["read-only-endpoints"]
+                if read_only_endpoints.strip() == "":
+                    continue
+                for ep in read_only_endpoints.split(","):
+                    roe.add(ep)
+        except json.JSONDecodeError:
+            raise ValueError("Relation data are not valid JSON.")
+
+    return roe
+
+
+def get_read_only_endpoint_hostnames(relation_data: list) -> List[str]:
+    """Returns the read-only-endpoint hostnames from the relation data.
+
+    Args:
+        relation_data: The dictionary that contains the info
+    Returns:
+        a set that contains the read-only-endpoint hostnames
+    """
+    roe = get_read_only_endpoint(relation_data)
+    roe_hostnames = []
+    for r in roe:
+        if ":" in r:
+            roe_hostnames.append(r.split(":")[0])
+        else:
+            raise ValueError("Malformed endpoint")
+    return roe_hostnames
+
+
+async def remove_leader_unit(ops_test: OpsTest, application_name: str):
+    """Removes the leader unit of a specified application.
+
+    Args:
+        ops_test: The ops test framework instance
+        application_name: The name of the application
+    """
+    leader_unit = None
+    for app_unit in ops_test.model.applications[application_name].units:
+        is_leader = await app_unit.is_leader_from_status()
+        if is_leader:
+            leader_unit = app_unit.name
+
+    units_to_destroy = [leader_unit]
+
+    for unit_to_destroy in units_to_destroy:
+        await ops_test.model.destroy_units(unit_to_destroy)
+
+    count = len(ops_test.model.applications[application_name].units)
+
+    application = ops_test.model.applications[application_name]
+    await ops_test.model.block_until(lambda: len(application.units) == count)
+
+    if count > 0:
+        await ops_test.model.wait_for_idle(
+            apps=[application_name],
+            status="active",
+            raise_on_blocked=True,
+            timeout=1000,
+        )
+
+
+async def get_unit_hostname(ops_test: OpsTest, app_name: str) -> List[str]:
+    """Retrieves hostnames of given application units.
+
+    Args:
+        ops_test: The ops test framework instance
+        app_name: The name of the application
+    Returns:
+        a list that contains the hostnames of a given application
+    """
+    units = [app_unit.name for app_unit in ops_test.model.applications[app_name].units]
+    status = await ops_test.model.get_status()  # noqa: F821
```

Comment:
why the `noqa` here?

Reply:
There was a problem with lint on the zookeeper charm. I checked that there is no problem in the charm so I will remove it.